Add gatkToWdlWrapper

knoblett opened the pull request:
Follows issue broadinstitute/dsde-docs#1223

knoblett commented:
Still working on the test script, but I wanted to start this pull request where I could put files.

knoblett commented:
Test script done--no idea what the issue was with the first go-around but rewriting from scratch was apparently the way to go.

There is one finicky bit about the scripts I've uploaded: they assume that there is a folder inside the directory of jsons called `/WDLtools/`. All logic on the internet says that when using the `w` specification when opening the file to write, all folders not present will be created. So, if I say that I want to write a file called `.../WDLtools/mytool.wdl`, and the folder `WDLtools` didn't exist, it would create the folder, then the file. However, when I run it that way, python throws an error saying that `WDLtools` doesn't exist. My workaround is to just create an empty `WDLtools` folder inside the directory of jsons. I spent about half a day trying to fix the error, but I couldn't figure it out and figured this workaround was easier.

vdauwera commented:
Nice, good idea to make the tester script.

knoblett commented:
All WDL task files added. This is ready for review @vdauwera

vdauwera commented on scripts/wrappers/gatk/gatkToWdlWrapper.py:
```diff
@@ -0,0 +1,210 @@
+
+        #Write header and declare task
+        wdlFile.write("# --------------------------------------------------------------------------------------------\n" +
+                      "# The following " + toolName + " script was generated on " + datetime.date.today().strftime("%m/%d/%y") +
```
script -> wrapper here and on line 96

vdauwera replied:
Also would do "The following" -> "This"

vdauwera commented on scripts/wrappers/gatk/gatkToWdlWrapper.py:
```diff
@@ -0,0 +1,210 @@
+                      "# The following " + toolName + " script was generated on " + datetime.date.today().strftime("%m/%d/%y") +
+                      " for use with GATK version " + version + "\n" +
+                      "# For more information on using this script, please see the WDL repository at \n" +
+                      "# https://github.com/broadinstitute/wdl/tree/develop/scripts/libraries/gatk/README.md\n" +
```
Update link to readme (wrappers, not libraries, right?)

vdauwera commented on scripts/wrappers/gatk/gatkToWdlWrapper.py:
```diff
@@ -0,0 +1,210 @@
+                    wdlFile.write("\t" + getWdlType(cmdArgs[getArgIndexByName(engineData[toolName][b][0], cmdArgs)]["type"]) +
+                                  " ? " + checkName(engineData[toolName][b][0][2:]) + "\n")
+                else:
+                    print("There was an error determining the optional status of an engine argument: Input recieved: "
```
recieved -> received

vdauwera commented on scripts/wrappers/gatk/gatkToWdlWrapper.py:
```diff
@@ -0,0 +1,210 @@
+                        wdlFile.write("\t\t\t" + engineData[toolName][b][0] + " ${default=\"" + checkQuotes(engineData[toolName][b][1]) + "\" " +
+                                      checkName(engineData[toolName][b][0][2:]) + "} \\\n")
+                else:
+                    print("There was an error determining the optional status of an engine argument: Input recieved: " +
```
recieved -> received

vdauwera commented on scripts/wrappers/gatk/gatkToWdlWrapper.py:
```diff
@@ -0,0 +1,210 @@
+                      "\t}\n")
+
+        #write runtime
+        #TODO Replace docker image with public one
```
The docker below is public, isn't it? Do you mean replace it with a GATK-only docker once we have that? (which I should ping Brad on)

vdauwera commented on scripts/wrappers/gatk/testWDLTasks.py:
```diff
@@ -0,0 +1,21 @@
+import sys
+import os
+
+#To use: pass in the directory containing all generated WDL tasks as the first argument
```
You could make this shorter and unambiguous by giving an example command, eg

    testWDLTasks.py path/to/wdl_tasks_dir /path/to/wdltool.jar

vdauwera commented on scripts/wrappers/gatk/testWDLTasks.py:
```diff
@@ -0,0 +1,21 @@
+# This script will print out the name of the file as it checks the file.
+# If no errors are found, there will be an empty line under the filename.
+# If errors are found, the error message will be output under the filename.
+# This script should be run prior to uploading any auto-generated WDL tasks to the repository
```
Sentence missing a final period.

vdauwera commented:
Looks great! Just a few nitpicks.

vdauwera commented on scripts/wrappers/gatk/gatkToWdlWrapper.py:
```diff
@@ -0,0 +1,210 @@
+            else:
+                print("There was an error determining optional status of an argument. Input received: " +
+                      toolArgs[b]["required"] + " for the argument " + toolArgs[b]["synonyms"])
+        wdlFile.write("\t}\n")
```
Just noticed the commands all have a backslash on the last line. Did you test that that doesn't cause problems? Otherwise you might need to add some logic to strip it out before writing the closing brace.

knoblett commented:
I've addressed all your comments, updated the general readme (in /scripts), and added a readme specific to /wrappers/gatk. I'm wondering if I should create a sub-folder called 'WDLTasks_3.6' to put all of the task scripts into, but I'm not sure. It would certainly be useful later on when we generate tasks for other GATK versions.

I'm also wondering if I should upload the engine_args_per_tool.json file, as it is needed to run the gatkToWdlWrapper.py script.

vdauwera commented:
Ah, good idea on making a WDLTasks_3.6 subfolder.  You can include engine_args_per_tool.json  in that subfolder as well, yes.

vdauwera commented:
After that, feel free to rebase so we can do final checks before merging. Almost done!
